conn2res/reservoir.py

```diff
@@ -9,8 +9,223 @@
+
+        elif node_set == 'random':
+            # nodes we want to select from
+            nodes_from = np.setdiff1d(nodes_from, nodes_without)
+
+            # select random nodes
+            selected_nodes = np.random.choice(nodes_from, size=n_nodes)
+
+        elif node_set == 'shortest_path':
+            # calculate shortest paths between all nodes
+            D = distance_bin(self.w)
+            D = np.triu(D)  # remove repetitions
+
+            # nodes we want to select from
+            nodes_from = np.setdiff1d(nodes_from, nodes_without)
+
+            # shortest paths between all nodes of interest
+            D = D[np.ix_(nodes_from, nodes_from)]
+
+            # select all node pairs with requested shortest path from each other
+            if isinstance(kwargs['shortest_path'], str):
+                node_pairs = np.argwhere(D == np.amax(D))
+            elif isinstance(kwargs['shortest_path'], int):
+                node_pairs = np.argwhere(D == kwargs['shortest_path'])
+
+            # select requested number of nodes from the set above
+            if len(np.unique(node_pairs)) >= n_nodes:
+                i = 1
+                while len(np.unique(node_pairs[:i, :])) < n_nodes:
+                    i += 1
+                selected_nodes = nodes_from[np.unique(node_pairs[:i, :])]
+            else:
+                raise ValueError(
+                    'n_nodes do not exist with given shortest_path')
+
+        else:
+            # nodes we want to select from
+            nodes_from = np.setdiff1d(nodes_from, nodes_without)
+
+            # load resting-state networks and filter to active nodes
+            rsn_mapping = load_file('rsn_mapping.npy')
+            rsn_mapping = rsn_mapping[self.idx_node]
+
+            # get modules
+            module_ids, modules = get_modules(rsn_mapping)
+
+            if node_set in module_ids:
+                # select all nodes in the requested module
+                selected_nodes = [e for i, e in enumerate(
+                    modules) if (module_ids == node_set)[i]][0]
+
+                # intersection of nodes we want to select from
+                selected_nodes = np.intersect1d(selected_nodes, nodes_from)
+            else:
+                raise ValueError('node_set does not exist with given value')
+
+        return selected_nodes
+
+    def _update_component(self):
+        """
+        Updates a set of nodes so that they belong to one connected component
+
+        #TODO
+        """
+
+        # make sure that the connectivity matrix consists of one large component
+        comps, comp_sizes = get_components(self.w)
+        idx_node = comps == np.where(comp_sizes != 1)[0] + 1
+
+        # update node attributes
+        self._update_nodes(idx_node)
+
+    def _update_nodes(self, idx_node):
+        """
+        Updates attributes of a new subset of nodes
+
+        #TODO
+        """
+
+        if isinstance(idx_node, np.ndarray) and idx_node.dtype == np.bool:
```

Comment:
We keep self.n_nodes and self.idx_node in sync and we also reduce self.w to submatrix where idx_node is True.